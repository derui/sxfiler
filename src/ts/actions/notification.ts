import { Notification } from "../domains/notification";
import { AppAction } from "./type";

export enum ActionTypes {
  timeout = "notification_timeout",
  notify = "notification_notify",
  remove = "notification_remove",
}

type TimeoutAction = AppAction<ActionTypes.timeout, { notificationId: string }>;
type NotifyAction = AppAction<ActionTypes.notify, { notification: Notification }>;
type RemoveAction = AppAction<ActionTypes.remove, { notificationId: string }>;

export type Actions = TimeoutAction | NotifyAction | RemoveAction;

const timeout = (notificationId: string): TimeoutAction => {
  return { type: ActionTypes.timeout, notificationId };
};

const notify = (notification: Notification): NotifyAction => {
  return { type: ActionTypes.notify, notification };
};

const remove = (notificationId: string): RemoveAction => {
  return { type: ActionTypes.remove, notificationId };
};

export const actions = { timeout, notify, remove };
