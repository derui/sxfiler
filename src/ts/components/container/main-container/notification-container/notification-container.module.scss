@import "../../../ui/commom-style/_colors.scss";
@import "../../../ui/commom-style/_variables.scss";
@import "../../../ui/commom-style/reset.module.css";

.root {
  position: relative;

  height: 100%;
  width: 100%;
  max-width: 100%;

  border-top: 1px solid $sol_base2;
}
