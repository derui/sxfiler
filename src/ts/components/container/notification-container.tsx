import * as React from "react";

interface Prop {}

const NotificationContainer: React.FC<Prop> = props => {
  return null;
};

export default NotificationContainer;
