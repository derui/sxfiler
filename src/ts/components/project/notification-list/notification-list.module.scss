
$messageWidth: 150px;

.root {
  composes: list-base from "../../ui/list/list.module.scss";

  position: absolute;
  flex-direction: column;
  width: $messageWidth;
}
