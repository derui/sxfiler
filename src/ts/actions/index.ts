import { Actions as NotificationActions } from "./notification";
import { Actions as FilerActions } from "./filer";
import { Actions as KeymapActions } from "./key-map";
import { Actions as TaskActions } from "./task";
import { Actions as HistoryActions } from "./history";
import { Actions as FinderAction } from "./finder";

export type Actions = NotificationActions | FilerActions | KeymapActions | TaskActions | HistoryActions | FinderAction;

export { ActionTypes } from "./type";
