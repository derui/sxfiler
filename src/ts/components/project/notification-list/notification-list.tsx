import * as React from "react";
import { Notification } from "../../domains/notification";
import NotificationItem, { TimeoutCallback } from "./notification-item";

interface Prop {
  // notifications given from server
  notifications: Notification[];
  // list of timeouted notification id
  timeouts: string[];
  // callback to timeout item
  onItemTimeouted: TimeoutCallback;
}

// make item
function toComponent(notification: Notification, timeouts: string[], onItemTimeouted: TimeoutCallback) {
  const timeouted = timeouts.find(v => v === notification.id) !== undefined;
  return <NotificationItem item={notification} timeouted={timeouted} onItemTimeouted={onItemTimeouted} />;
}

const NotificationList: React.FC<Prop> = prop => {
  const { notifications, timeouts, onItemTimeouted } = prop;

  const components = notifications.map(v => toComponent(v, timeouts, onItemTimeouted));

  return <ul className="fp-NotificationList">{components}</ul>;
};

export default NotificationList;
