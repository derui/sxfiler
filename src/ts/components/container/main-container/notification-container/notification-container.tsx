import * as React from "react";

import * as Element from "../../../ui/element/element";
import { ContextLike } from "../../../../context";
import { State } from "../../../../states/notification";
import { createUseCase } from "../../../../usecases/notification/timeout";
import { Component as NotificationList } from "../../../project/notification-list/notification-list";
import LocatorContext from "../../../../locator";

// eslint-disable-next-line @typescript-eslint/no-var-requires
const styles = require("./notification-container.module.scss");

export interface Props {
  state: State;
}

export class Component extends React.PureComponent<Props> {
  public render() {
    const { state } = this.props;
    const timeouts = state.timeouts.messages.map(v => v.id);

    return (
      <LocatorContext.Consumer>
        {({ context }) => {
          if (!context) {
            return null;
          }

          return (
            <Element.Component className={styles.root}>
              <NotificationList
                notifications={state.notifications.messages}
                timeouts={timeouts}
                onNotificationHidden={this.handleNotificationHidden(context)}
              />
            </Element.Component>
          );
        }}
      </LocatorContext.Consumer>
    );
  }

  private handleNotificationHidden = (context: ContextLike) => (id: string) => {
    context.execute(createUseCase(), {
      notificationId: id,
    });
  };
}
