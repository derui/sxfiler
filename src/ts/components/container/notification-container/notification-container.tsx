import * as React from "react";
import { Context } from "../../../context";
import { State } from "../../../types/store-state/notification";
import TimeoutUseCase from "../../../usecases/notification/timeout";
import { TimeoutCallback } from "../../project/notification-item/notification-item";
import * as NotificationList from "../../project/notification-list/notification-list";

// tslint:disable-next-line
const styles = require("./notification-container.module.scss");

export interface Props {
  context: Context;
  state: State;
}

export class Component extends React.PureComponent<Props> {
  public render() {
    const { state } = this.props;

    return (
      <div className={styles.root}>
        <NotificationList.Conmonent
          notifications={state.notifications.items}
          onItemTimeouted={this.handleItemTimeouted}
        />
      </div>
    );
  }
  private handleItemTimeouted: TimeoutCallback = (id: string) => {
    this.props.context.execute(new TimeoutUseCase(), {
      notificationId: id,
    });
  };
}
