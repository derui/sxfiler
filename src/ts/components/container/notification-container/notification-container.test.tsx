import { shallow } from "enzyme";
import * as React from "react";
import renderer from "react-test-renderer";

import { Actions } from "../../../actions/notification";
import { createMessage, createProgress, Level } from "../../../domains/notification";
import Notifications from "../../../domains/notifications";
import { UseCaseLike } from "../../../usecases/type";
import * as NotificationList from "../../project/notification-list/notification-list";
import { Component } from "./notification-container";

const context = (done?: (arg: any) => void) => ({
  execute<P>(usecase: UseCaseLike<Actions, P>, arg: P) {
    if (done) {
      done(arg);
    }
  },
});

describe("Container", () => {
  describe("Notification Container", () => {
    it("should render correctly", () => {
      const state = {
        notifications: new Notifications([
          createMessage("message", Level.Info, "message"),
          createProgress("progress", Level.Info, {
            process: "process",
            target: 100,
            current: 10,
          }),
        ]),
        timeouts: new Notifications(),
      };
      const tree = renderer.create(<Component context={context()} state={state} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it("should execute usecase with hidden notification id", () => {
      const state = {
        notifications: new Notifications([
          createMessage("message", Level.Info, "message"),
          createProgress("progress", Level.Info, {
            process: "process",
            target: 100,
            current: 10,
          }),
        ]),
        timeouts: new Notifications(),
      };
      const stub = (arg: any) => {
        expect(arg).toEqual({ notificationId: "id" });
      };
      const wrapper = shallow(<Component context={context(stub)} state={state} />);
      wrapper.find(NotificationList.Component).simulate("notificationHidden", "id");
    });
  });
});
