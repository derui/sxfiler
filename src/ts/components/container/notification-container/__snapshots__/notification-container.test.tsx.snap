// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container Notification Container should render correctly 1`] = `
<div
  className="root"
>
  <div
    className="root"
    role="list"
  >
    <div
      aria-selected={false}
      className="root"
      data-level="info"
      role="listitem"
    >
      message
    </div>
  </div>
</div>
`;
